### ISSUE TITLE
Spaces ignored in define

### ISSUE THREAD

**Issue by packadal** [fastbuild/fastbuild#59]
The CMake tests try to pass lots of strange characters by define, and the following fails:

-DTARGETSTRING="\"hello`~!@^$*)(+-=}{][:^'.?/  ,&|^^<>world;\""

Notice that there are two spaces after the forward slash.
Only the first space in a sequence is kept in this situation.

This could be simplified to: 
-DFOO="\"                  \""

In the resulting source file, FOO would be " " instead of "                  ".

It seems the fastbuild tokenizer does not recognize the escape double quote and simplifies the whitespace when it should not.

Thanks

**Comment by inbilla** [comment]
I can try attempt a fix for this if you like?

**Comment by ffulin** [comment]
I suspect I know where the problem lies, but if I'm right there are some subtleties that will need to be considered to implement a fix. Might be best to wait until I can take a look so I can confirm what's needed for the fix?

**Comment by ffulin** [comment]
The tokenization in AString doesn't respect the \ escape character. It would be relatively easy to add an optional escape character to that function, which would make it parse your examples correctly, but there are some subtleties.

**a) User Error**
A really common user error with the MSVC compiler is for people to do this:
/I"path\path\"
Which will cause all following commands to be considered as part of the include path, causing really hard to understand errors.
FASTBuild detects and silently fixes this case right now (ObjectNode::BuildFullArgs) by checking for an include /I option which is terminated this way, and removing the escaping backslash at the end.
If we fix the parsing, we'll need to move this fix to the parsing of the bff, instead of post-tokenization where it is now.

**b) Non-obvious case**
In tests I did before (based on comments I left in the code), I saw:
```C++
            /I\"dir1\" /I\"dir2\"
```
is two tokens, while
```C++
            /I"dir1\" /I"dir2"
```
is considered one token.

I had to think about this for a while, and I think it makes sense, but it means we can't simply skip escaped quotes. We'll need to be careful with the implementation, by treating escaped quotes as non-quote characters.

The good news is that there are tests for AString::Tokenize which we can easily extend to verify all scenarios.

**Comment by fran6co** [comment]
Related to this issue `%1` and `%2` can't be escaped

**Comment by ffulin** [comment]
Various arg parsing fixes have been implemented in https://github.com/fastbuild/fastbuild/commit/36a8b4dbce16c930afcd14210c80a69529ec3ada

It's unclear if those fixes alone are enough to resolve the specific issue reported here (some further testing will be required).

**Comment by ffulin** [comment]
I've tested this with the latest fixes and the spaces seem to be handled correctly, so this will be resolved as of the next release v1.13. I'll keep this issue open until that version is released.

Note that quoted args that also contains quotes require args to be escaped:
```
"-DSTRING=\"thing with spaces\""
```
or
```
-DSTRING="\"thing with spaces\""
```

**Comment by ffulin** [comment]
v1.13 has now been released and includes this fix. Thanks again for reporting the bug.
